app/controllers/concerns/authentication.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+# Concern for handling authentication.
+# Note that this concern is based on the generated code.
+module Authentication
+  extend ActiveSupport::Concern
+
+  # Apache is configured so that:
+  # - /webauth/login requires a shibboleth authenticated user. Thus, redirecting a user to it triggers login.
+  # - /Shibboleth.sso/Logout logs the user out of shibboleth.
+  # - /queues is limited to members of the sdr:developer group.
+  # - Other pages do not require authentication.
+  # - It provides the following request headers:
+  #  - X-Remote-User
+  #  - X-Groups
+  #  - X-Person-Name (first name)
+  #  - X-Person-Formal-Name (full name)
+
+  REMOTE_USER_HEADER = 'X-Remote-User'
+  GROUPS_HEADER = 'X-Groups'
+  FIRST_NAME_HEADER = 'X-Person-Name'
+  NAME_HEADER = 'X-Person-Formal-Name'
+
+  # The development environment does not have Shibboleth authentication.
+  # Instead, it will use this hardcoded user.
+  DEV_REMOTE_USER = 'a.user@stanford.edu'
+  DEV_FIRST_NAME = 'A.'
+  DEV_NAME = 'A. User'
+
+  SHIBBOLETH_LOGOUT_PATH = '/Shibboleth.sso/Logout'
+
+  included do
+    # authentication will be called before require_authentication.
+    # It will authenticate the user if there is a user.
+    # This will be called for all controller actions.
+    # require_authentication will also be called for all controller actions,
+    # unless skipped with allow_unauthenticated_access.
+    before_action :authentication, :require_authentication
+    helper_method :authenticated?, :current_user
+  end
+
+  class_methods do
+    def allow_unauthenticated_access(**options)
+      skip_before_action :require_authentication, **options
+    end
+  end
+
+  def current_user
+    Current.user
+  end
+
+  private
+
+  def remote_user
+    return DEV_REMOTE_USER if Rails.env.development?
+
+    request.headers['X-Remote-User']
```

Comment:
Could this use the `REMOTE_USER_HEADER` constant instead of the string?

Reply:
Yes, fixed.